# Data Wrangling and Analyzing

In this project, we'll gather, assess, and clean some Tweeter data then act on it through analysis, visualization and/or modeling.

## Table of Contents
- [Gather](#gather)
- [Access](#access)
- [Clean](#clean)
- [Analysis](#analysis)
- [Conclusion](#conclusion)

<a id='gather'></a>
## Gather

In [32]:
# Import necessary libraries
import pandas as pd
import requests
import tweepy

Load `twitter-archive-enhanced.csv` file.

In [89]:
df_tweets = pd.read_csv('twitter-archive-enhanced.csv')

Download `image_predictions.tsv` file using requests.

In [23]:
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
r = requests.get(url)

In [24]:
# Check status of response
r.status_code

200

In [31]:
# Write response to file
with open('image_predictions.tsv', 'w+') as file:
    file.write(r.text)    

In [144]:
df_images = pd.read_csv('image_predictions.tsv', sep='\t')

Gather tweeter information with tweepy.

In [3]:
# Connect to tweepy
with open('keys.txt', 'r') as file:
    api_key = file.readline().split()[2]
    api_secret = file.readline().split()[2]
    access_token = file.readline().split()[2]
    access_secret = file.readline().split()[2]
    
auth = tweepy.OAuthHandler(api_key, api_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit = True)

In [57]:
# Download tweet data and save to tweet_json.txt
import json

file = open('tweet_json.txt', 'w+')
count = 0
for tweet_id in df['tweet_id']:
    count = count + 1
    print('{count} / 2356, {id}'.format(count=count, id=tweet_id))
    try:
        file.write(json.dumps(api.get_status(id=tweet_id)._json) + '\n')
    except:
        print('Deleted: ', tweet_id)
        continue
file.close()

1 / 2356, 892420643555336193
2 / 2356, 892177421306343426
3 / 2356, 891815181378084864
4 / 2356, 891689557279858688
5 / 2356, 891327558926688256
6 / 2356, 891087950875897856
7 / 2356, 890971913173991426
8 / 2356, 890729181411237888
9 / 2356, 890609185150312448
10 / 2356, 890240255349198849
11 / 2356, 890006608113172480
12 / 2356, 889880896479866881
13 / 2356, 889665388333682689
14 / 2356, 889638837579907072
15 / 2356, 889531135344209921
16 / 2356, 889278841981685760
17 / 2356, 888917238123831296
18 / 2356, 888804989199671297
19 / 2356, 888554962724278272
20 / 2356, 888202515573088257
Deleted:  888202515573088257
21 / 2356, 888078434458587136
22 / 2356, 887705289381826560
23 / 2356, 887517139158093824
24 / 2356, 887473957103951883
25 / 2356, 887343217045368832
26 / 2356, 887101392804085760
27 / 2356, 886983233522544640
28 / 2356, 886736880519319552
29 / 2356, 886680336477933568
30 / 2356, 886366144734445568
31 / 2356, 886267009285017600
32 / 2356, 886258384151887873
33 / 2356, 886054160

259 / 2356, 843604394117681152
260 / 2356, 843235543001513987
261 / 2356, 842892208864923648
Deleted:  842892208864923648
262 / 2356, 842846295480000512
263 / 2356, 842765311967449089
264 / 2356, 842535590457499648
265 / 2356, 842163532590374912
266 / 2356, 842115215311396866
267 / 2356, 841833993020538882
268 / 2356, 841680585030541313
269 / 2356, 841439858740625411
270 / 2356, 841320156043304961
271 / 2356, 841314665196081154
272 / 2356, 841077006473256960
273 / 2356, 840761248237133825
274 / 2356, 840728873075638272
275 / 2356, 840698636975636481
276 / 2356, 840696689258311684
277 / 2356, 840632337062862849
278 / 2356, 840370681858686976
279 / 2356, 840268004936019968
280 / 2356, 839990271299457024
281 / 2356, 839549326359670784
282 / 2356, 839290600511926273
283 / 2356, 839239871831150596
284 / 2356, 838952994649550848
285 / 2356, 838921590096166913
286 / 2356, 838916489579200512
287 / 2356, 838831947270979586
288 / 2356, 838561493054533637
289 / 2356, 838476387338051585
290 / 2356

518 / 2356, 810896069567610880
519 / 2356, 810657578271330305
520 / 2356, 810284430598270976
521 / 2356, 810254108431155201
522 / 2356, 809920764300447744
523 / 2356, 809808892968534016
524 / 2356, 809448704142938112
525 / 2356, 809220051211603969
526 / 2356, 809084759137812480
527 / 2356, 808838249661788160
528 / 2356, 808733504066486276
529 / 2356, 808501579447930884
530 / 2356, 808344865868283904
531 / 2356, 808134635716833280
532 / 2356, 808106460588765185
533 / 2356, 808001312164028416
534 / 2356, 807621403335917568
535 / 2356, 807106840509214720
536 / 2356, 807059379405148160
537 / 2356, 807010152071229440
538 / 2356, 806629075125202948
539 / 2356, 806620845233815552
540 / 2356, 806576416489959424
541 / 2356, 806542213899489280
542 / 2356, 806242860592926720
543 / 2356, 806219024703037440
544 / 2356, 805958939288408065
545 / 2356, 805932879469572096
546 / 2356, 805826884734976000
547 / 2356, 805823200554876929
548 / 2356, 805520635690676224
549 / 2356, 805487436403003392
550 / 23

781 / 2356, 775733305207554048
782 / 2356, 775729183532220416
783 / 2356, 775364825476165632
784 / 2356, 775350846108426240
785 / 2356, 775096608509886464
Deleted:  775096608509886464
786 / 2356, 775085132600442880
787 / 2356, 774757898236878852
788 / 2356, 774639387460112384
789 / 2356, 774314403806253056
790 / 2356, 773985732834758656
791 / 2356, 773922284943896577
792 / 2356, 773704687002451968
793 / 2356, 773670353721753600
794 / 2356, 773547596996571136
795 / 2356, 773336787167145985
796 / 2356, 773308824254029826
797 / 2356, 773247561583001600
798 / 2356, 773191612633579521
799 / 2356, 772877495989305348
800 / 2356, 772826264096874500
801 / 2356, 772615324260794368
802 / 2356, 772581559778025472
803 / 2356, 772193107915964416
804 / 2356, 772152991789019136
805 / 2356, 772117678702071809
806 / 2356, 772114945936949249
807 / 2356, 772102971039580160
808 / 2356, 771908950375665664
809 / 2356, 771770456517009408
810 / 2356, 771500966810099713
811 / 2356, 771380798096281600
812 / 2356

1040 / 2356, 744234799360020481
1041 / 2356, 744223424764059648
1042 / 2356, 743980027717509120
1043 / 2356, 743895849529389061
1044 / 2356, 743835915802583040
1045 / 2356, 743609206067040256
1046 / 2356, 743595368194129920
1047 / 2356, 743545585370791937
1048 / 2356, 743510151680958465
1049 / 2356, 743253157753532416
1050 / 2356, 743222593470234624
1051 / 2356, 743210557239623680
1052 / 2356, 742534281772302336
1053 / 2356, 742528092657332225
1054 / 2356, 742465774154047488
1055 / 2356, 742423170473463808
1056 / 2356, 742385895052087300
1057 / 2356, 742161199639494656
1058 / 2356, 742150209887731712
1059 / 2356, 741793263812808706
1060 / 2356, 741743634094141440
1061 / 2356, 741438259667034112
1062 / 2356, 741303864243200000
1063 / 2356, 741099773336379392
1064 / 2356, 741067306818797568
1065 / 2356, 740995100998766593
1066 / 2356, 740711788199743490
1067 / 2356, 740699697422163968
1068 / 2356, 740676976021798912
1069 / 2356, 740373189193256964
1070 / 2356, 740365076218183684
1071 / 2

1298 / 2356, 707776935007539200
1299 / 2356, 707741517457260545
1300 / 2356, 707738799544082433
1301 / 2356, 707693576495472641
1302 / 2356, 707629649552134146
1303 / 2356, 707610948723478529
1304 / 2356, 707420581654872064
1305 / 2356, 707411934438625280
1306 / 2356, 707387676719185920
1307 / 2356, 707377100785885184
1308 / 2356, 707315916783140866
1309 / 2356, 707297311098011648
1310 / 2356, 707059547140169728
1311 / 2356, 707038192327901184
1312 / 2356, 707021089608753152
1313 / 2356, 707014260413456384
1314 / 2356, 706904523814649856
1315 / 2356, 706901761596989440
1316 / 2356, 706681918348251136
1317 / 2356, 706644897839910912
1318 / 2356, 706593038911545345
1319 / 2356, 706538006853918722
1320 / 2356, 706516534877929472
1321 / 2356, 706346369204748288
1322 / 2356, 706310011488698368
1323 / 2356, 706291001778950144
1324 / 2356, 706265994973601792
1325 / 2356, 706169069255446529
1326 / 2356, 706166467411222528
1327 / 2356, 706153300320784384
1328 / 2356, 705975130514706432
1329 / 2

1556 / 2356, 688894073864884227
1557 / 2356, 688828561667567616
1558 / 2356, 688804835492233216
1559 / 2356, 688789766343622656
1560 / 2356, 688547210804498433
1561 / 2356, 688519176466644993
1562 / 2356, 688385280030670848
1563 / 2356, 688211956440801280
1564 / 2356, 688179443353796608
1565 / 2356, 688116655151435777
1566 / 2356, 688064179421470721
1567 / 2356, 687841446767013888
1568 / 2356, 687826841265172480
1569 / 2356, 687818504314159109
1570 / 2356, 687807801670897665
1571 / 2356, 687732144991551489
1572 / 2356, 687704180304273409
1573 / 2356, 687664829264453632
1574 / 2356, 687494652870668288
1575 / 2356, 687480748861947905
1576 / 2356, 687476254459715584
1577 / 2356, 687460506001633280
1578 / 2356, 687399393394311168
1579 / 2356, 687317306314240000
1580 / 2356, 687312378585812992
1581 / 2356, 687127927494963200
1582 / 2356, 687124485711986689
1583 / 2356, 687109925361856513
1584 / 2356, 687102708889812993
1585 / 2356, 687096057537363968
1586 / 2356, 686947101016735744
1587 / 2

1812 / 2356, 676819651066732545
1813 / 2356, 676811746707918848
1814 / 2356, 676776431406465024
1815 / 2356, 676617503762681856
1816 / 2356, 676613908052996102
1817 / 2356, 676606785097199616
1818 / 2356, 676603393314578432
1819 / 2356, 676593408224403456
1820 / 2356, 676590572941893632
1821 / 2356, 676588346097852417
1822 / 2356, 676582956622721024
1823 / 2356, 676575501977128964
1824 / 2356, 676533798876651520
1825 / 2356, 676496375194980353
1826 / 2356, 676470639084101634
1827 / 2356, 676440007570247681
1828 / 2356, 676430933382295552
1829 / 2356, 676263575653122048
1830 / 2356, 676237365392908289
1831 / 2356, 676219687039057920
1832 / 2356, 676215927814406144
1833 / 2356, 676191832485810177
1834 / 2356, 676146341966438401
1835 / 2356, 676121918416756736
1836 / 2356, 676101918813499392
1837 / 2356, 676098748976615425
1838 / 2356, 676089483918516224
1839 / 2356, 675898130735476737
1840 / 2356, 675891555769696257
1841 / 2356, 675888385639251968
1842 / 2356, 675878199931371520
1843 / 2

2070 / 2356, 671134062904504320
2071 / 2356, 671122204919246848
2072 / 2356, 671115716440031232
2073 / 2356, 671109016219725825
2074 / 2356, 670995969505435648
2075 / 2356, 670842764863651840
2076 / 2356, 670840546554966016
2077 / 2356, 670838202509447168
2078 / 2356, 670833812859932673
2079 / 2356, 670832455012716544
2080 / 2356, 670826280409919488
2081 / 2356, 670823764196741120
2082 / 2356, 670822709593571328
2083 / 2356, 670815497391357952
2084 / 2356, 670811965569282048
2085 / 2356, 670807719151067136
2086 / 2356, 670804601705242624
2087 / 2356, 670803562457407488
2088 / 2356, 670797304698376195
2089 / 2356, 670792680469889025
2090 / 2356, 670789397210615808
2091 / 2356, 670786190031921152
2092 / 2356, 670783437142401025
2093 / 2356, 670782429121134593
2094 / 2356, 670780561024270336
2095 / 2356, 670778058496974848
2096 / 2356, 670764103623966721
2097 / 2356, 670755717859713024
2098 / 2356, 670733412878163972
2099 / 2356, 670727704916926465
2100 / 2356, 670717338665226240
2101 / 2

2327 / 2356, 666411507551481857
2328 / 2356, 666407126856765440
2329 / 2356, 666396247373291520
2330 / 2356, 666373753744588802
2331 / 2356, 666362758909284353
2332 / 2356, 666353288456101888
2333 / 2356, 666345417576210432
2334 / 2356, 666337882303524864
2335 / 2356, 666293911632134144
2336 / 2356, 666287406224695296
2337 / 2356, 666273097616637952
2338 / 2356, 666268910803644416
2339 / 2356, 666104133288665088
2340 / 2356, 666102155909144576
2341 / 2356, 666099513787052032
2342 / 2356, 666094000022159362
2343 / 2356, 666082916733198337
2344 / 2356, 666073100786774016
2345 / 2356, 666071193221509120
2346 / 2356, 666063827256086533
2347 / 2356, 666058600524156928
2348 / 2356, 666057090499244032
2349 / 2356, 666055525042405380
2350 / 2356, 666051853826850816
2351 / 2356, 666050758794694657
2352 / 2356, 666049248165822465
2353 / 2356, 666044226329800704
2354 / 2356, 666033412701032449
2355 / 2356, 666029285002620928
2356 / 2356, 666020888022790149


In [83]:
# Extract retweet count and favorite count of each tweet
likes = []
with open('tweet_json.txt', 'r') as file:
    for line in file:
        tweet = json.loads(line)
        tweet_id = tweet['id_str']
        retweet = tweet['retweet_count'] 
        favorite = tweet['favorite_count']
        
        likes.append({'tweet_id': tweet_id,
                        'retweet_count': retweet,
                        'favorite_count':favorite})
df_likes = pd.DataFrame(likes, columns = ['tweet_id', 'retweet_count', 'favorite_count'])

<a id='access'></a>
## Access

Now we have 3 files dataframes on hand, `df_tweets`, `df_images`, and `df_likes`.

In [92]:
df_tweets.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [179]:
df_tweets.sample(3)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
247  845459076796616705                    NaN                  NaN   
468  817056546584727552                    NaN                  NaN   
43   884162670584377345                    NaN                  NaN   

                     timestamp  \
247  2017-03-25 02:15:26 +0000   
468  2017-01-05 17:13:55 +0000   
43   2017-07-09 21:29:42 +0000   

                                                source  \
247  <a href="http://twitter.com/download/iphone" r...   
468  <a href="http://twitter.com/download/iphone" r...   
43   <a href="http://twitter.com/download/iphone" r...   

                                                  text  retweeted_status_id  \
247  RT @dog_rates: Here's a heartwarming scene of ...         7.562885e+17   
468  This is Chloe. She fell asleep at the wheel. A...                  NaN   
43   Meet Yogi. He doesn't have any important dog m...                  NaN   

     retweeted_status_user_id retweeted_status_timestamp  \
247              4.196984e+09  2016-07-22 00:43:32 +0000   
468                       NaN                        NaN   
43                        NaN                        NaN   

                                         expanded_urls  rating_numerator  \
247  https://twitter.com/dog_rates/status/756288534...                12   
468  https://twitter.com/dog_rates/status/817056546...                11   
43   https://twitter.com/dog_rates/status/884162670...                12   

     rating_denominator   name  doggo floofer pupper puppo  
247                  10   None   None    None   None  None  
468                  10  Chloe   None    None   None  None  
43                   10   Yogi  doggo    None   None  None

In [99]:
df_tweets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [116]:
df_tweets.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [150]:
sum(df_tweets.tweet_id.duplicated())

0

In [137]:
df_tweets.rating_numerator.sort_values()

315        0
1016       0
2335       1
2261       1
2338       1
        ... 
2074     420
188      420
189      666
313      960
979     1776
Name: rating_numerator, Length: 2356, dtype: int64

In [140]:
df_tweets.query('rating_numerator == 1776')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
979  749981277374128128                    NaN                  NaN   

                     timestamp  \
979  2016-07-04 15:00:45 +0000   

                                                source  \
979  <a href="https://about.twitter.com/products/tw...   

                                                  text  retweeted_status_id  \
979  This is Atticus. He's quite simply America af....                  NaN   

     retweeted_status_user_id retweeted_status_timestamp  \
979                       NaN                        NaN   

                                         expanded_urls  rating_numerator  \
979  https://twitter.com/dog_rates/status/749981277...              1776   

     rating_denominator     name doggo floofer pupper puppo  
979                  10  Atticus  None    None   None  None

In [162]:
df_tweets.rating_denominator.sort_values()

313       0
2335      2
516       7
1576     10
1575     10
       ... 
1635    110
1779    120
1634    130
902     150
1120    170
Name: rating_denominator, Length: 2356, dtype: int64

In [171]:
df_tweets.query('rating_denominator == 0 or rating_denominator == 2 or rating_denominator == 7')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
313   835246439529840640           8.352460e+17           26259576.0   
516   810984652412424192                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   

                      timestamp  \
313   2017-02-24 21:54:03 +0000   
516   2016-12-19 23:06:23 +0000   
2335  2015-11-16 16:11:11 +0000   

                                                 source  \
313   <a href="http://twitter.com/download/iphone" r...   
516   <a href="http://twitter.com/download/iphone" r...   
2335  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
313   @jonnysun @Lin_Manuel ok jomny I know you're e...                  NaN   
516   Meet Sam. She smiles 24/7 &amp; secretly aspir...                  NaN   
2335  This is an Albanian 3 1/2 legged  Episcopalian...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
313                        NaN                        NaN   
516                        NaN                        NaN   
2335                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
313                                                 NaN               960   
516   https://www.gofundme.com/sams-smile,https://tw...                24   
2335  https://twitter.com/dog_rates/status/666287406...                 1   

      rating_denominator  name doggo floofer pupper puppo  
313                    0  None  None    None   None  None  
516                    7   Sam  None    None   None  None  
2335                   2    an  None    None   None  None

In [177]:
df_tweets[df_tweets.expanded_urls.isnull()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
30    886267009285017600           8.862664e+17         2.281182e+09   
55    881633300179243008           8.816070e+17         4.738443e+07   
64    879674319642796034           8.795538e+17         3.105441e+09   
113   870726314365509632           8.707262e+17         1.648776e+07   
148   863427515083354112           8.634256e+17         7.759620e+07   
179   857214891891077121           8.571567e+17         1.806710e+08   
185   856330835276025856                    NaN                  NaN   
186   856288084350160898           8.562860e+17         2.792810e+08   
188   855862651834028034           8.558616e+17         1.943518e+08   
189   855860136149123072           8.558585e+17         1.361572e+07   
218   850333567704068097           8.503288e+17         2.195506e+07   
228   848213670039564288           8.482121e+17         4.196984e+09   
234   847617282490613760           8.476062e+17         4.196984e+09   
274   840698636975636481           8.406983e+17         8.405479e+17   
290   838150277551247360           8.381455e+17         2.195506e+07   
291   838085839343206401           8.380855e+17         2.894131e+09   
313   835246439529840640           8.352460e+17         2.625958e+07   
342   832088576586297345           8.320875e+17         3.058208e+07   
346   831926988323639298           8.319030e+17         2.068372e+07   
375   828361771580813312                    NaN                  NaN   
387   826598799820865537           8.265984e+17         4.196984e+09   
409   823333489516937216           8.233264e+17         1.582854e+09   
427   821153421864615936           8.211526e+17         1.132119e+08   
498   813130366689148928           8.131273e+17         4.196984e+09   
513   811647686436880384           8.116272e+17         4.196984e+09   
570   801854953262350336           8.018543e+17         1.185634e+07   
576   800859414831898624           8.008580e+17         2.918590e+08   
611   797165961484890113           7.971238e+17         2.916630e+07   
701   786051337297522688           7.727430e+17         7.305050e+17   
707   785515384317313025                    NaN                  NaN   
843   766714921925144576           7.667118e+17         4.196984e+09   
857   763956972077010945           7.638652e+17         1.584641e+07   
967   750381685133418496           7.501805e+17         4.717297e+09   
1005  747651430853525504           7.476487e+17         4.196984e+09   
1080  738891149612572673           7.384119e+17         3.589728e+08   
1295  707983188426153984           7.079801e+17         2.319108e+09   
1345  704491224099647488           7.044857e+17         2.878549e+07   
1445  696518437233913856                    NaN                  NaN   
1446  696490539101908992           6.964887e+17         4.196984e+09   
1474  693644216740769793           6.936422e+17         4.196984e+09   
1479  693582294167244802           6.935722e+17         1.198989e+09   
1497  692423280028966913           6.924173e+17         4.196984e+09   
1523  690607260360429569           6.903413e+17         4.670367e+08   
1598  686035780142297088           6.860340e+17         4.196984e+09   
1605  685681090388975616           6.855479e+17         4.196984e+09   
1618  684969860808454144           6.849598e+17         4.196984e+09   
1663  682808988178739200           6.827884e+17         4.196984e+09   
1689  681340665377193984           6.813394e+17         4.196984e+09   
1774  678023323247357953           6.780211e+17         4.196984e+09   
1819  676590572941893632           6.765883e+17         4.196984e+09   
1844  675849018447167488           6.758457e+17         4.196984e+09   
1895  674742531037511680           6.747400e+17         4.196984e+09   
1905  674606911342424069           6.744689e+17         4.196984e+09   
1914  674330906434379776           6.658147e+17         1.637468e+07   
1940  673716320723169284           6.737159e+17         

In [178]:
df_tweets[df_tweets.name == "None"]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
5     891087950875897856                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
...                  ...                    ...                  ...   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2351  666049248165822465                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
5     2017-07-29 00:08:17 +0000   
7     2017-07-28 00:22:40 +0000   
12    2017-07-25 01:55:32 +0000   
24    2017-07-18 16:08:03 +0000   
25    2017-07-18 00:07:08 +0000   
...                         ...   
2342  2015-11-16 02:38:37 +0000   
2343  2015-11-16 01:59:36 +0000   
2344  2015-11-16 01:52:02 +0000   
2351  2015-11-16 00:24:50 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
5     <a href="http://twitter.com/download/iphone" r...   
7     <a href="http://twitter.com/download/iphone" r...   
12    <a href="http://twitter.com/download/iphone" r...   
24    <a href="http://twitter.com/download/iphone" r...   
25    <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2342  <a href="http://twitter.com/download/iphone" r...   
2343  <a href="http://twitter.com/download/iphone" r...   
2344  <a href="http://twitter.com/download/iphone" r...   
2351  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
5     Here we have a majestic great white breaching ...                  NaN   
7     When you watch your owner call another dog a g...                  NaN   
12    Here's a puppo that seems to be on the fence a...                  NaN   
24    You may not have known you needed to see this ...                  NaN   
25    This... is a Jubilant Antarctic House Bear. We...                  NaN   
...                                                 ...                  ...   
2342  Here we have a well-established sunblockerspan...                  NaN   
2343  Let's hope this flight isn't Malaysian (lol). ...                  NaN   
2344  Here we have a northern speckled Rhododendron....                  NaN   
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
5                          NaN                        NaN   
7                          NaN                        NaN   
12                         NaN                        NaN   
24                         NaN                        NaN   
25                         NaN                        NaN   
...                        ...                        ...   
2342                       NaN                        NaN   
2343                       NaN                        NaN   
2344                       NaN                        NaN   
2351                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
5     https://twitter.com/dog_rates/status/891087950...                13   
7     https://twitter.com/dog_rates/status/890729181...                13   
12    https://twitter.com/dog_rates/status/889665388...                13   
24    https://twitt

In [145]:
df_images.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [146]:
df_images.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [180]:
df_images.sample(3)

tweet_id                                          jpg_url  \
1358  760656994973933572  https://pbs.twimg.com/media/Co5lf-KW8AAIwJw.jpg   
1643  808001312164028416  https://pbs.twimg.com/media/CzaY5UdUoAAC91S.jpg   
293   671347597085433856  https://pbs.twimg.com/media/CVEbFDRWsAAkN_7.jpg   

      img_num                  p1   p1_conf  p1_dog  \
1358        1    golden_retriever  0.760546    True   
1643        1  Labrador_retriever  0.730959    True   
293         1        picket_fence  0.382918   False   

                             p2   p2_conf  p2_dog  \
1358         Labrador_retriever  0.232079    True   
1643  Staffordshire_bullterrier  0.130726    True   
293                 rain_barrel  0.108809   False   

                                  p3   p3_conf  p3_dog  
1358                         redbone  0.002874    True  
1643  American_Staffordshire_terrier  0.028853    True  
293                      plastic_bag  0.038878   False

In [151]:
sum(df_images.tweet_id.duplicated())

0

In [152]:
df_images.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [153]:
df_likes.head()

tweet_id  retweet_count  favorite_count
0  892420643555336193           7305           34812
1  892177421306343426           5457           30164
2  891815181378084864           3598           22706
3  891689557279858688           7490           38065
4  891327558926688256           8051           36342

In [154]:
sum(df_likes.duplicated())

0

In [155]:
df_likes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2331 entries, 0 to 2330
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   tweet_id        2331 non-null   object
 1   retweet_count   2331 non-null   int64 
 2   favorite_count  2331 non-null   int64 
dtypes: int64(2), object(1)
memory usage: 54.8+ KB


In [156]:
df_likes.describe()

retweet_count  favorite_count
count    2331.000000     2331.000000
mean     2569.824539     7269.266409
std      4346.939571    11291.352969
min         1.000000        0.000000
25%       521.500000     1258.000000
50%      1197.000000     3149.000000
75%      2975.000000     8885.500000
max     73712.000000   149541.000000

### Quality

##### `df_tweets` table
- ~~Erroneous datatypes (tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, retweeted_status_id, retweeted_status_user_id, retweet_status_timestamp columns)~~
- ~~Missing information on in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id, retweeted_status_user_id, and retweet_status_timestamp~~
- Rating numerators having very large range, some are not actual ratings (e.g. 1776 is not rating for a dog but the year of Declaration of Independence of the United States)
- ~~745 dogs are named "None", some are named "a"~~
- ~~rating_numerator and rating_denominator are wrong for tweet_id '666287406224695296', it should be 9/10 instead of 1/2~~
- rating_numerator and rating_denominator are wrong for tweet_id '810984652412424192', 24/7 is not a rating. There is no rating for this dog
- ~~Some of the tweets are not original tweets~~
- ~~Some of the tweets contain videos instead of images~~

##### `df_images` table
- ~~Erroneous datatype (tweet_id)~~
- ~~Mix use of upper and lower cases first letters of prediction (p1, p2, p3)~~

##### `df_likes` table
- ~~Number of entries does not match number of entries of `df_tweets` - there are some deleted tweets~~
- ~~Erroneous datatype (retweet_count, favorite_count)~~

### Tidiness

##### `df_tweets` table
- ~~Four columns of dog stages~~
- ~~Retweet counts and favorite counts should be part of the `df_tweets` table~~

<a id='clean'></a>
## Clean

In [707]:
tweets_clean = df_tweets.copy()
images_clean = df_images.copy()
likes_clean = df_likes.copy()

### Quality

#### `tweets`: Some of the tweets are not original tweets

##### Define

Since we only care about original tweets with images, we are dropping all the retweets.

##### Code

In [708]:
# Make a mask of all tweets with 'in_reply_to_user_id' (meaning it's a reply), 
# 'retweeted_status_id' (meaning it's a retweet), and withough 'expanded_urls'
# (meaning it does not contain images)

mask = (tweets_clean['in_reply_to_user_id'].notnull()) | (tweets_clean['retweeted_status_id'].notnull())

In [709]:
tweets_clean.drop(tweets_clean[mask].index, axis = 0, inplace=True)

In [710]:
# Remove in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id, retweeted_status_user_id, 
# retweeted_status_timestamp because these rows have been emptied
tweets_clean.drop(columns=['in_reply_to_status_id', 'in_reply_to_user_id', 'retweeted_status_id', 
                           'retweeted_status_user_id', 'retweeted_status_timestamp'], inplace=True)

##### Test

In [711]:
tweets_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   tweet_id            2097 non-null   int64 
 1   timestamp           2097 non-null   object
 2   source              2097 non-null   object
 3   text                2097 non-null   object
 4   expanded_urls       2094 non-null   object
 5   rating_numerator    2097 non-null   int64 
 6   rating_denominator  2097 non-null   int64 
 7   name                2097 non-null   object
 8   doggo               2097 non-null   object
 9   floofer             2097 non-null   object
 10  pupper              2097 non-null   object
 11  puppo               2097 non-null   object
dtypes: int64(3), object(9)
memory usage: 213.0+ KB


#### `tweets`: Some of the tweets contain videos instead of images

##### Define

Remove all tweets without an expanded_urls or expanded_urls contains the word 'vine'

##### Code

In [712]:
# Find all the expanded_urls that are either null or contains the word "vine"
mask = (tweets_clean['expanded_urls'].isnull()) | (tweets_clean['expanded_urls'].str.contains('vine', regex=False))

In [713]:
tweets_clean.drop(tweets_clean[mask].index, axis = 0, inplace=True)

##### Test

In [714]:
tweets_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2003 entries, 0 to 2355
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   tweet_id            2003 non-null   int64 
 1   timestamp           2003 non-null   object
 2   source              2003 non-null   object
 3   text                2003 non-null   object
 4   expanded_urls       2003 non-null   object
 5   rating_numerator    2003 non-null   int64 
 6   rating_denominator  2003 non-null   int64 
 7   name                2003 non-null   object
 8   doggo               2003 non-null   object
 9   floofer             2003 non-null   object
 10  pupper              2003 non-null   object
 11  puppo               2003 non-null   object
dtypes: int64(3), object(9)
memory usage: 203.4+ KB


In [715]:
tweets_clean[tweets_clean['expanded_urls'].str.contains('vine')]

Empty DataFrame
Columns: [tweet_id, timestamp, source, text, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [716]:
tweets_clean[tweets_clean['expanded_urls'].isnull()]

Empty DataFrame
Columns: [tweet_id, timestamp, source, text, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

#### `tweets`: 745 dogs are named "None", some are named "a"

##### Define

Change dog names from "None"/"a" to empty string.

##### Code

In [768]:
tweets_clean['name'].replace('None', '', inplace=True)
tweets_clean['name'].replace('a', '', inplace=True)

##### Test

In [781]:
tweets_clean['name'].value_counts()

             609
Charlie       11
Lucy          10
Cooper        10
Oliver        10
            ... 
Pip            1
Sojourner      1
Dug            1
Pluto          1
Stewie         1
Name: name, Length: 934, dtype: int64

In [771]:
sum(tweets_clean['name'].isnull())

0

#### `tweets`: rating_numerator and rating_denominator are wrong for tweet_id '666287406224695296', it should be 9/10 instead of 1/2

##### Define

Change `rating_numerator` to 9 and `rating_denominator` to 10 for tweet_id `666287406224695296`

##### Code

In [776]:
tweets_clean.loc[tweets_clean.tweet_id == '666287406224695296', 'rating_numerator'] = 9
tweets_clean.loc[tweets_clean.tweet_id == '666287406224695296', 'rating_denominator'] = 10

##### Test

In [777]:
tweets_clean[tweets_clean.tweet_id == '666287406224695296']

tweet_id                 timestamp  \
1982  666287406224695296 2015-11-16 16:11:11+00:00   

                                                 source  \
1982  <a href="http://twitter.com/download/iphone" r...   

                                                   text  \
1982  This is an Albanian 3 1/2 legged  Episcopalian...   

                                          expanded_urls  rating_numerator  \
1982  https://twitter.com/dog_rates/status/666287406...                 9   

      rating_denominator name dog_stages  retweet_count  favorite_count  
1982                  10   an                      57.0           130.0

#### `tweets_clean`: rating_numerator and rating_denominator are wrong for tweet_id '810984652412424192', 24/7 is not a rating. There is no rating for this dog

##### Define

Remove entry because it does not have ratings.

##### Code

In [ ]:
tweets_clean.drop(tweets_clean[tweets_clean.tweet_id == '666287406224695296'], inplace=True)

#### `tweets_clean`: Erroneous datatypes (tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, retweeted_status_id, retweeted_status_user_id, retweet_status_timestamp columns)

##### Define

Change `tweet_id` to string, change `timestamp` to datetime

##### Code

In [795]:
tweets_clean.tweet_id = tweets_clean.tweet_id.astype(str)
tweets_clean.timestamp = pd.to_datetime(tweets_clean.timestamp)

##### Test

In [734]:
tweets_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2003 entries, 0 to 2355
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   tweet_id            2003 non-null   object             
 1   timestamp           2003 non-null   datetime64[ns, UTC]
 2   source              2003 non-null   object             
 3   text                2003 non-null   object             
 4   expanded_urls       2003 non-null   object             
 5   rating_numerator    2003 non-null   int64              
 6   rating_denominator  2003 non-null   int64              
 7   name                2003 non-null   object             
 8   dog_stages          2003 non-null   object             
dtypes: datetime64[ns, UTC](1), int64(2), object(6)
memory usage: 156.5+ KB


#### `images_clean`: Erroneous datatype (tweet_id)

##### Define

Change `tweet_id` to string

##### Code

In [739]:
images_clean.tweet_id = images_clean.tweet_id.astype(str)

##### Test

In [740]:
images_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   object 
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(1), object(5)
memory usage: 152.1+ KB


#### `tweets_clean`: Erroneous datatype (retweet_count, favorite_count)

##### Define

Change `retweet_count` and `favorite_count` to int

##### Code

In [793]:
tweets_clean.retweet_count = tweets_clean.retweet_count.astype(int)
tweets_clean.favorite_count = tweets_clean.favorite_count.astype(int)

##### Test

In [794]:
tweets_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2003 entries, 0 to 2002
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   tweet_id            2003 non-null   object             
 1   timestamp           2003 non-null   datetime64[ns, UTC]
 2   source              2003 non-null   object             
 3   text                2003 non-null   object             
 4   expanded_urls       2003 non-null   object             
 5   rating_numerator    2003 non-null   int64              
 6   rating_denominator  2003 non-null   int64              
 7   name                2003 non-null   object             
 8   dog_stages          2003 non-null   object             
 9   retweet_count       2003 non-null   int32              
 10  favorite_count      2003 non-null   int32              
dtypes: datetime64[ns, UTC](1), int32(2), int64(2), object(6)
memory usage: 172.1+ KB


#### `images_clean`: Mix use of upper and lower cases first letters of prediction (p1, p2, p3)

##### Define

Change every string in `p1`, `p2`, `p3` to lower case

##### Code

In [756]:
images_clean['p1'] = images_clean['p1'].str.lower()
images_clean['p2'] = images_clean['p2'].str.lower()
images_clean['p3'] = images_clean['p3'].str.lower()

##### Test

In [763]:
images_clean.p1.str.islower().value_counts()

True    2075
Name: p1, dtype: int64

In [764]:
images_clean.p2.str.islower().value_counts()

True    2075
Name: p2, dtype: int64

In [765]:
images_clean.p3.str.islower().value_counts()

True    2075
Name: p3, dtype: int64

### Tidiness

#### `tweets_clean`: Four columns of dog stages

##### Define

Merge the `doggo`, `floofer`, `pupper`, and `puppo` columns to a `dog_stages` column.


##### Code

In [717]:
# Replace 'None' and NaN with an empty string for columns 'doggo', 
# 'floofer', 'pupper' and 'puppo'

import numpy as np

tweets_clean['doggo'].replace('None', '', inplace=True)
tweets_clean['doggo'].replace(np.NaN, '', inplace=True)
tweets_clean['floofer'].replace('None', '', inplace=True)
tweets_clean['floofer'].replace(np.NaN, '', inplace=True)
tweets_clean['pupper'].replace('None', '', inplace=True)
tweets_clean['pupper'].replace(np.NaN, '', inplace=True)
tweets_clean['puppo'].replace('None', '', inplace=True)
tweets_clean['puppo'].replace(np.NaN, '', inplace=True)

In [718]:
# Extract information from tweets_clean['text'] to see if dog stage is mentioned
tweets_clean['dog_stages'] = tweets_clean['text'].str.extract('(doggo|floofer|pupper|puppo)', expand = True)

In [719]:
tweets_clean['dog_stages'].value_counts()

pupper     223
doggo       73
puppo       28
floofer      3
Name: dog_stages, dtype: int64

In [720]:
# Combine dog stages - some dogs have multiple stages
tweets_clean['dog_stages'] = tweets_clean['doggo'] + tweets_clean['floofer'] + tweets_clean['pupper'] + tweets_clean['puppo']
tweets_clean.loc[tweets_clean['dog_stages'] == 'doggopupper', 'dog_stages'] = 'doggo, pupper'
tweets_clean.loc[tweets_clean['dog_stages'] == 'doggopuppo', 'dog_stages'] = 'doggo, puppo'
tweets_clean.loc[tweets_clean['dog_stages'] == 'doggofloofer', 'dog_stages'] = 'doggo, floofer'

In [721]:
# Drop the original 'doggo','floofer','pupper','puppo' columns
tweets_clean.drop(['doggo','floofer','pupper','puppo'], axis=1, inplace = True)

##### Test

In [722]:
tweets_clean['dog_stages'].value_counts()

                  1694
pupper             204
doggo               66
puppo               22
doggo, pupper        8
floofer              7
doggo, floofer       1
doggo, puppo         1
Name: dog_stages, dtype: int64

#### `likes_clean`: Retweet counts and favorite counts should be part of the `tweets_clean` table

##### Define

Merge `likes_clean` and `tweets_clean`


##### Code

In [735]:
tweets_clean = pd.merge(tweets_clean, likes_clean, how='left', on='tweet_id')

In [791]:
# Null entries for retweet and favorite counts are due to deleted tweets
tweets_clean['retweet_count'] = tweets_clean['retweet_count'].fillna(0)
tweets_clean['favorite_count'] = tweets_clean['favorite_count'].fillna(0)

##### Test

In [792]:
tweets_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2003 entries, 0 to 2002
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   tweet_id            2003 non-null   object             
 1   timestamp           2003 non-null   datetime64[ns, UTC]
 2   source              2003 non-null   object             
 3   text                2003 non-null   object             
 4   expanded_urls       2003 non-null   object             
 5   rating_numerator    2003 non-null   int64              
 6   rating_denominator  2003 non-null   int64              
 7   name                2003 non-null   object             
 8   dog_stages          2003 non-null   object             
 9   retweet_count       2003 non-null   float64            
 10  favorite_count      2003 non-null   float64            
dtypes: datetime64[ns, UTC](1), float64(2), int64(2), object(6)
memory usage: 187.8+ KB


<a id='analysis'></a>
## Analysis

<a id='conclusion'></a>
## Conclusion